lib/libbackscrub.cc

```diff
@@ -99,94 +122,125 @@ static normalization_t get_normalization(modeltype_t type) {
+	if (nullptr == ctx.interpreter) {
+		_dbg(ctx, "error: unable to build model interpreter\n");
+		bs_maskgen_delete(pctx);
+		return nullptr;
+	}
 
 	// Allocate tensor buffers.
-	TFLITE_MINIMAL_CHECK(ctx.interpreter->AllocateTensors() == kTfLiteOk);
+	if (ctx.interpreter->AllocateTensors() != kTfLiteOk) {
+		_dbg(ctx, "error: unable to allocate tensor buffers\n");
+		bs_maskgen_delete(pctx);
+		return nullptr;
+	}
 
 	// set interpreter params
-	ctx.interpreter->SetNumThreads(info.threads);
+	ctx.interpreter->SetNumThreads(threads);
 	ctx.interpreter->SetAllowFp16PrecisionForFp32(true);
 
 	// get input and output tensor as cv::Mat
-	info.input = getTensorMat(info, ctx.interpreter->inputs ()[0]);
-	info.output = getTensorMat(info, ctx.interpreter->outputs()[0]);
-	info.ratio = (float)info.input.cols/(float) info.input.rows;
+	ctx.input = getTensorMat(ctx, ctx.interpreter->inputs ()[0]);
+	ctx.output = getTensorMat(ctx, ctx.interpreter->outputs()[0]);
+	ctx.ratio = (float)ctx.input.cols/(float) ctx.input.rows;
 
-	// initialize mask and square ROI in center
-	info.roidim = cv::Rect((info.width-info.height/info.ratio)/2,0,info.height/info.ratio,info.height);
-	info.mask = cv::Mat::ones(info.height,info.width,CV_8UC1)*255;
-	info.mroi = info.mask(info.roidim);
+	// initialize mask and model-aspect ROI in center
+	ctx.roidim = cv::Rect((width-height/ctx.ratio)/2,0,height/ctx.ratio,height);
+	ctx.mask = cv::Mat::ones(height,width,CV_8UC1)*255;
+	ctx.mroi = ctx.mask(ctx.roidim);
 
 	// mask blurring size
-	info.blur = cv::Size(5,5);
+	ctx.blur = cv::Size(5,5);
 
 	// create Mat for small mask
-	info.ofinal = cv::Mat(info.output.rows,info.output.cols,CV_8UC1);
-	return 1;
+	ctx.ofinal = cv::Mat(ctx.output.rows,ctx.output.cols,CV_8UC1);
+	return pctx;
 }
 
-void drop_tensorflow(calcinfo_t &info) {
-	if (info.debug)
-		_dbg(info, "dropping tensorflow context\n");
+void bs_maskgen_delete(void *context) {
+	if (nullptr == context)
+		return;
+	backscrub_ctx_t &ctx = *((backscrub_ctx_t *)context);
 	// clear all mask data
-	info.ofinal.deallocate();
-	info.mask.deallocate();
-	info.input.deallocate();
-	info.output.deallocate();
-	// clear internal context if present
-	if (info.backscrub_ctx) {
-		backscrub_ctx_t &ctx = *((backscrub_ctx_t *)info.backscrub_ctx);
-		// drop interpreter
+	ctx.ofinal.deallocate();
+	ctx.mask.deallocate();
+	ctx.input.deallocate();
+	ctx.output.deallocate();
+	// drop interpreter (if present)
+	if (ctx.interpreter != nullptr)
 		ctx.interpreter.reset();
-		// drop model
+	// drop model (if present)
+	if (ctx.model != nullptr)
 		ctx.model.reset();
-		// drop context
-		delete &ctx;
-		info.backscrub_ctx = nullptr;
-	}
+	delete &ctx;
 }
 
-int calc_mask(calcinfo_t &info) {
-	// Ensure we have a context from init_tensorflow()
-	TFLITE_MINIMAL_CHECK(info.backscrub_ctx!=NULL);
-	backscrub_ctx_t &ctx = *((backscrub_ctx_t *)info.backscrub_ctx);
+int bs_maskgen_process(void *context, cv::Mat &frame, cv::Mat &mask) {
```

Comment:
Should return `bool` instead if the return value is just used as an success indicator.

Reply:
Agreed, may need to be revisited as/when a C API appears (which would need to use C99 `stdbool.h` include)